### PR TITLE
GitHub issue handler

### PR DESCRIPTION
**Pull request by jhaigh0**
This is the main pr for "automatic github issue creation" of the Mantid error reporter improvements Epic.

What's been added:

- A new database model `GithubIssue` comprised of a repo address and an issue number
- `github_issue_manager.py` which contains the main funcion `get_or_create_github_issue` which does the following
  - Searches the database for a githubIssue with a matching repo and stacktrace (the stacktrace is compared irrespective of the install location / any local informaiton)
    - If one is found, a comment containing the report info is added to that issue
    - If no issue is found, then a new one is made

Two new environment variables have been added
 - `GIT_AUTH_TOKEN` is the auth token of the account used to create the github issues / comments (on the live version of the error reports site this will be an auth token from https://github.com/Mantid-Error-Reporter)
 - `GIT_ISSUE_REPO` the repository to create the issues in (e.g `'ISISNeutronMuon/mantid-isis-support'`)

I've also changed the slack message poster so that, id an issue is created / found, it'll be linked in the slack message.

### CODE REVIEW

File: web/services/github_issue_manager/test_trim_stacktrace.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+    def test_user_specific_dirs_are_removed(self):
+        username = "my_cool_user_name"
+        test_trace = f'File "/home/{username}/mantidworkbench/lib/python3.8/site-packages/mantidqt/widgets/memorywidget/memoryview.py", line 98, in _set_value'\
+                      '   @Slot(int, float, float)'\
+                      'KeyboardInterrupt'
+        self.assertNotIn(username, _trim_stacktrace(test_trace))
+
+    def test_result_is_independant_of_path_style(self):
+        pass
```
Comment (robertapplin) [inline web/services/github_issue_manager/test_trim_stacktrace.py:15]: Is this meant to be here?

File: web/services/github_issue_manager/github_issue_manager.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+Mantid version: $version
+OS: $os
+
+**Additionl Information**
+$info
+
+**Stack trace**
+```$stacktrace```
+""")
+comment_text_template = Template("""
```
Comment (robertapplin) [inline web/services/github_issue_manager/github_issue_manager.py:26]: Same for these constant variables in caps?

File: web/services/github_issue_manager/github_issue_manager.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        return None
+    trimmed_trace = _trim_stacktrace(trace)
+    for raw_trace, github_issue in ErrorReport.objects.exclude(
+            githubIssue__isnull=True).values_list('stacktrace', 'githubIssue'):
+        if _trim_stacktrace(raw_trace) == trimmed_trace:
+            return GithubIssue.objects.get(id=github_issue)
+    return None
+
+
+def _search_for_repeat_user(uid: str, github_issue: GithubIssue) -> bool:
```
Comment (robertapplin) [inline web/services/github_issue_manager/github_issue_manager.py:151]: Can this function using the python `any()` function?

File: web/services/github_issue_manager/github_issue_manager.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+                                  labels=[error_report_label],
+                                  body=issue_text)
+        logger.info(f'Created issue {issue.url})')
+        return GithubIssue.objects.create(repoName=issue_repo,
+                                          issueNumber=issue.number)
+
+
+def _trim_stacktrace(stacktrace: str) -> str:
+    """
+    Returns a rimmed and os non-specific version of the stacktrace given
```
Comment (robertapplin) [inline web/services/github_issue_manager/github_issue_manager.py:111]:
```suggestion
    Returns a trimmed and os non-specific version of the stacktrace given
```

File: web/services/github_issue_manager/github_issue_manager.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+line_exp = re.compile(r"\s*File \".*(mantidqt|mantidqtinterfaces|"
+                      r"workbench|scripts)(\/|\\)(.*)(\", line \d+, in \S+)")
+issue_text_template = Template("""
+Name: $name
+Email: $email
+
+Mantid version: $version
+OS: $os
+
+**Additionl Information**
```
Comment (robertapplin) [inline web/services/github_issue_manager/github_issue_manager.py:20]:
```suggestion
**Additional Information**
```

File: web/services/github_issue_manager/github_issue_manager.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+```$stacktrace```
+""")
+comment_text_template = Template("""
+Name: $name
+Email: $email
+
+Mantid version: $version
+OS: $os
+
+**Additionl Information**
```
Comment (robertapplin) [inline web/services/github_issue_manager/github_issue_manager.py:33]:
```suggestion
**Additional Information**
```

File: web/services/github_issue_manager/github_issue_manager.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    with the report's key information. If not, create a new issue.
+
+    Return None in the following cases:
+     - There is no stack trace and no additional information in the report
+     - A GIT_AUTH_TOKEN has not been set
+     - The bug has already been submitted by the user (identified via the uid)
+       and they have not left any additional information
+
+    Args:
+        report: The report recived by ErrorViewSet
```
Comment (robertapplin) [inline web/services/github_issue_manager/github_issue_manager.py:51]:
```suggestion
        report: The report recieved by ErrorViewSet
```

**Review by robertapplin — CHANGES_REQUESTED** [review]

File: web/services/github_issue_manager/test_search_for_matching_stacktrace.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
+from django.test import TestCase
+from services.models import ErrorReport, GithubIssue
+from services.github_issue_manager.github_issue_manager import _search_for_matching_stacktrace
+
+
+class MatchingStackTraceSearchTest(TestCase):
+    entries = [
```
Comment (thomashampson) [inline web/services/github_issue_manager/test_search_for_matching_stacktrace.py:7]: I think the following would not be matched by the regex:

`  File "/opt/mantidworkbench6.9/lib/python3.10/site-packages/mantid/simpleapi.py", line 1083, in __call__`

and also the following kind of thing that I found in the error reports slack channel:

```
Exception: MDWorkspace::getExperimentInfo(): expInfoIndex is out of range.
  at line 152 in '/usr/local/anaconda/envs/mantid-dev/plugins/python/algorithms/ConvertWANDSCDtoQ.py'
```

File: web/services/github_issue_manager/test_trim_stacktrace.py [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+    def test_user_specific_dirs_are_removed(self):
+        username = "my_cool_user_name"
+        test_trace = f'File "/home/{username}/mantidworkbench/lib/python3.8/site-packages/mantidqt/widgets/memorywidget/memoryview.py", line 98, in _set_value'\
+                      '   @Slot(int, float, float)'\
+                      'KeyboardInterrupt'
+        self.assertNotIn(username, _trim_stacktrace(test_trace))
+
+    def test_line_trimmer_file_lines(self):
+        examples = {
```
Comment (thomashampson) [inline web/services/github_issue_manager/test_trim_stacktrace.py:15]: you might want to add the above examples to this test too

**Review by thomashampson — CHANGES_REQUESTED** [review]
I think I found a couple of cases that the stack trace trimming would miss, which could result in new issues being created rather than comments added to existing issues.

**Review by thomashampson — APPROVED** [review]
Let's get this in and get it working

**Pull request merged**